src/contracts/libraries/GPv2SafeERC20.sol

```diff
@@ -29,36 +39,55 @@ library GPv2SafeERC20 {
-    /// Openzeppelin's SafeERC20. Unlike Openzeppelin's implementation, this
-    /// function does not revert if an address without code is called.
-    ///
-    /// @param token The token targeted by the call.
-    /// @param data The call data (encoded using abi.encode or one of its
-    /// variants).
-    function _callOptionalReturn(IERC20 token, bytes memory data) private {
-        // solhint-disable-next-line avoid-low-level-calls
-        (bool success, bytes memory returndata) = address(token).call(data);
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            let freeMemoryPointer := mload(0x40)
+            mstore(freeMemoryPointer, selector_)
+            mstore(add(freeMemoryPointer, 4), from)
+            mstore(add(freeMemoryPointer, 36), to)
+            mstore(add(freeMemoryPointer, 68), value)
 
-        if (!success) {
-            // Assembly used to revert with correctly encoded error message.
-            // solhint-disable-next-line no-inline-assembly
-            assembly {
-                revert(add(returndata, 0x20), mload(returndata))
+            if iszero(call(gas(), token, 0, freeMemoryPointer, 100, 0, 0)) {
+                returndatacopy(0, 0, returndatasize())
+                revert(0, returndatasize())
             }
         }
-        if (returndata.length > 0) {
-            // Return data is optional
-            require(
-                abi.decode(returndata, (bool)),
-                "GPv2SafeERC20: failed transfer"
-            );
+
+        require(getLastTansferResult(), "GPv2: failed transferFrom");
+    }
+
+    /// @dev Verifies that the last return was a successful `transfer*` call.
+    /// This is done by checking that the return data is either empty, or
+    /// is a valid ABI encoded boolean.
+    function getLastTansferResult() private pure returns (bool success) {
+        // NOTE: Inspecting previous return data requires assembly. Note that
+        // we write the return data to memory 0 in the case where the return
+        // data size is 32, this is OK since the first 64 bytes of memory are
+        // reserved by Solidy as a scratch space that can be used within
+        // assembly blocks.
+        // <https://docs.soliditylang.org/en/v0.7.6/internals/layout_in_memory.html>
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            switch returndatasize()
+                // Non-standard ERC20 transfer without return.
+                case 0 {
+                    success := 1
+                }
+                // Standard ERC20 transfer returning boolean success value.
+                case 32 {
+                    returndatacopy(0, 0, returndatasize())
+                    success := mload(0)
+                    if gt(success, 1) {
+                        // NOTE: Not a valid ABI encoded boolean, which must be
+                        // either 0 or 1.
+                        revert(0, 0)
+                    }
```

Comment:
Can we test that this is the same behaviour as currently? I'd be afraid of some clowny ERC20s encoding true as keccak("ERC20.return.true")  😉

Reply:
This is a very good point. Based on the Solidity documentation I had understood that booleans can only be 0 or 1, and not any non-zero value.

<https://docs.soliditylang.org/en/v0.7.6/types.html#booleans>
> bool: The possible values are constants true and false.

and
<https://docs.soliditylang.org/en/v0.7.6/abi-spec.html?highlight=bool%20assembly>
> bool: equivalent to uint8 restricted to the values 0 and 1. For computing the function selector, bool is used.

**However**, it appears that `abi.decode(...)` accepts values that are not 0 or 1. Consider the following code:
```solidity
// SPDX-License-Identifier: CC0-1.0
pragma solidity ^0.7.6;
contract Foo {
    function testBoolDecode(uint256 value) external pure returns (bool flag, uint256 marker) {
        bytes memory encoded = abi.encode(value);
        flag = abi.decode(encoded, (bool));
        if (flag) {
            marker = 1;
        }
    }
}
```

This code returns `(flag = false, marker = 0)` if `value == 0` and `(flag = true, marker = 1)` otherwise.

One more data point, if the code above is slightly modified:
```solidity
pragma solidity ^0.8.1;
```
Then the code reverts if `value > 1` (so it seems to be more in line with the linked documentation).

Reply:
**HOWEVER**, it looks like this is dependent on the ABI encoding pragma. If we modify the code sligtly:
```diff
-pragma abicoder v2;
```

Then this will return `(flag = true, maker = 1)` for any `value > 0`. So it behaves differently than the code :point_up:.

Reply:
So the question is which ABI encoding scheme do we want to support? (I suggest `abicoder v2` just because its going to be the default going forward, but also because it more closely follows the documentation).

Reply:
AFAIK abicoder v2 is enabled by default only from Solidity v0.8, so unless the token contracts we are going to use decided to opt in they would be relying on v1. I'd say to try to copy the behavior of SafeERC20, which is `v1` (no pragma and `require(abi.decode(returndata, (bool)))`).

Reply:
> so unless the token contracts we are going to use decided to opt in they would be relying on v1.

So, one thing to note is that even with `abicoder v1`, the encoded bools are 0 or 1:
```
// SPDX-License-Identifier: CC0-1.0
pragma solidity >=0.5.0;
contract Foo {
    function testBoolEncode(uint256 value) external pure returns (bool coercedValue, bytes memory encoded) {
        assembly {
            coercedValue := value
        }
        encoded = abi.encode(coercedValue);
    }
}
```

This returns `(coercedValue = true, encoded = 0x...01)` for all `value > 1` (I checked this back to Solidity 0.5.17). This is in line with the [documented ABI specification](https://docs.soliditylang.org/en/v0.7.6/abi-spec.html?highlight=bool#types). It just appears that the `abi.decode` is more permissive.

My worry for accepting non-standard return values is that the contract doesn't really know what they mean. So what if some weird ERC20 returns `keccak256("ERC20.return.false")` instead of `false` for a failed transfer? Since we are trying to wrap `transfer(...) -> bool` and `transfer(...)`, then I think we don't need to be permissive with the bool decoding, and a contract that implements `transfer(...) -> bytes32` should not be supported in general.

Reply:
As I see it, the question is if, in case that the returned value from an `ERC20` token is `v > 1`:
1. to return `true` (`abiencode v1`) (I assume from the discussion that this also holds for `v = 0x0100`.)
2. to revert (`abiencode v2`)

My argument for using `abiencode v1` is that `SafeERC20` is a de facto standard library and is known to work with all current token implementations that matter. If we go down the `v2` route then we are letting more transfers fail, that is we decide that tokens that do magic with the returned value would not be tradable on GPv2 but may be in e.g. Uniswap.
I actually believe there are no real tokens that do this, so maybe this discussion is too much. I guess we'll know for sure when the first popular exchange is implemented in Solidity v0.8, then any issue should become apparent.

Reply:
This is a good point, and I'm convinced! Lets keep it compatible with the existing `SafeERC20` implementation. I'll add a comment explaining the decision.

Reply:
Updated code to be compatible with OpenZeppelin `SafeERC20` WRT ABI decoding bool values.